### PR TITLE
Fix attention mask for RT-DETR-based models

### PR DESCRIPTION
**Pull request by materight**
# What does this PR do?

Currently the attention mask is not handled correctly for RT-DETR-based models (`rt_detr`, `rt_detr_v2`, `d_fine`). The docs for `get_contrastive_denoising_training_group` say that it should return a `torch.FloatTensor` for the attention mask, but it actually returns `torch.bool`. The mask is then just summed up with the attention weights, so it just adds 0 or 1. 

This PR changes `get_contrastive_denoising_training_group` to return a `torch.FloatTensor` with `-inf` on the masked weights, and adapts all the custom `MultiHeadAttention` to correctly do the conversion if a boolean mask is given in input,  as done in [`nn.MultiheadAttention`](https://github.com/pytorch/pytorch/blob/ba56102387ef21a3b04b357e5b183d48f0afefc7/torch/nn/modules/activation.py#L1233), where the boolean mask is [converted to a float tensor](https://github.com/pytorch/pytorch/blob/8f31aa97a3e1e17bed29b6cedf9884f0c6b145e9/torch/nn/functional.py#L6050).

Related to [this issue](https://github.com/Peterande/D-FINE/issues/276#issuecomment-3052369924).

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@qubvel @VladOS95-cyber

### CODE REVIEW

**Comment by qubvel** [comment]
run-slow: conditional_detr, d_fine, detr, maskformer, rt_detr

**Review by qubvel — APPROVED** [review]
@materight thanks for the fix!
